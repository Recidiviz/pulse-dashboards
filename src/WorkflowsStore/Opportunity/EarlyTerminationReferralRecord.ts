// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

type EarlyTerminationReason = {
  eligibleDate?: string;
  supervisionType?: string;
  revocationDate?: string;
  supervisionLevel?: string;
};

export type EarlyTerminationCriteria = {
  criteriaName: string;
  reason: EarlyTerminationReason;
};

export type EarlyTerminationReferralRecord = {
  stateCode: string;
  externalId: string;
  formInformation: {
    clientName: string;
    convictionCounty: string;
    judicialDistrictCode: string;
    criminalNumber: string;
    judgeName: string;
    priorCourtDate: string;
    sentenceLengthYears: string;
    crimeNames: string[];
    probationExpirationDate: string;
    probationOfficerFullName: string;
  };
  reasons: EarlyTerminationCriteria[];
  metadata: {
    multipleSentences: boolean;
    outOfState: boolean;
    ICOut: boolean;
  };
};

export interface TransformedEarlyTerminationReferral {
  stateCode: string;
  externalId: string;
  formInformation: {
    clientName: string;
    convictionCounty: string;
    judicialDistrictCode: string;
    criminalNumber: string;
    judgeName: string;
    priorCourtDate: Date;
    sentenceLengthYears: number;
    crimeNames: string[];
    probationExpirationDate: Date;
    probationOfficerFullName: string;
  };
  reasons: {
    pastEarlyDischarge?: { eligibleDate?: Date };
    eligibleSupervisionLevel?: { supervisionLevel?: string };
    eligibleSupervisionType?: { supervisionType?: string };
    notActiveRevocationStatus?: { revocationDate?: Date };
  };
  metadata: {
    multipleSentences: boolean;
    outOfState: boolean;
    ICOut: boolean;
  };
}

// Extends Record to facilitate `additionalDepositionLines` dynamic keys
export interface EarlyTerminationDraftData extends Record<string, string> {
  courtName: string;
  clientName: string;
  convictionCounty: string;
  plaintiff: string;
  finesAndFees: string;
  judicialDistrictCode: string;
  criminalNumber: string;
  judgeName: string;
  priorCourtDate: string;
  sentenceLengthYears: string;
  crimeNames: string;
  probationExpirationDate: string;
  probationOfficerFullName: string;
  statesAttorneyNumber: string;
  statesAttorneyPhoneNumber: string;
  statesAttorneyEmailAddress: string;
  statesAttorneyMailingAddress: string;
  sentenceLength: string;
}
