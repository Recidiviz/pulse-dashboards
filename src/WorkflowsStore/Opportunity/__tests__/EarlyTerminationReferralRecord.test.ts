// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { parseISO } from "date-fns";

import { transformReferral } from "../EarlyTerminationReferralRecord";

test("transform record", () => {
  const rawRecord = {
    stateCode: "US_OZ",
    externalId: "abc123",
    formInformation: {
      clientName: "Firstname Lastname",
      convictionCounty: "Bloom",
      judicialDistrictCode: "XX",
      criminalNumber: "123456",
      judgeName: "Judgename",
      priorCourtDate: "2022-01-01",
      sentenceLengthYears: "2",
      crimeNames: ["Crime1", "crime2"],
      probationExpirationDate: "2024-05-01",
      probationOfficerFullName: "Fakename Officername",
    },
    criteria: {
      supervisionEarlyDischargeDateWithin30Days: { eligibleDate: "2022-01-03" },
      usNdImpliedValidEarlyTerminationSupervisionLevel: {
        supervisionLevel: "MEDIUM",
      },
      usNdImpliedValidEarlyTerminationSentenceType: {
        supervisionType: "PROBATION",
      },
      usNdNotInActiveRevocationStatus: { revocationDate: undefined },
    },
    metadata: {
      multipleSentences: false,
      outOfState: false,
      ICOut: false,
    },
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});

test("revocation date may be defined", () => {
  const rawRecord = {
    stateCode: "US_OZ",
    externalId: "abc123",
    formInformation: {
      clientName: "Firstname Lastname",
      convictionCounty: "Bloom",
      judicialDistrictCode: "XX",
      criminalNumber: "123456",
      judgeName: "Judgename",
      priorCourtDate: "2022-01-01",
      sentenceLengthYears: "2",
      crimeNames: ["Crime1", "crime2"],
      probationExpirationDate: "2024-05-01",
      probationOfficerFullName: "Fakename Officername",
    },
    criteria: {
      supervisionEarlyDischargeDateWithin30Days: { eligibleDate: "2022-01-03" },
      usNdImpliedValidEarlyTerminationSupervisionLevel: {
        supervisionLevel: "MEDIUM",
      },
      usNdImpliedValidEarlyTerminationSentenceType: {
        supervisionType: "PROBATION",
      },
      usNdNotInActiveRevocationStatus: { revocationDate: "2021-11-13" },
    },
    metadata: {
      multipleSentences: false,
      outOfState: false,
      ICOut: false,
    },
  };

  expect(
    transformReferral(rawRecord)?.criteria.usNdNotInActiveRevocationStatus
      ?.revocationDate
  ).toEqual(parseISO("2021-11-13"));
});

test("fallback to legacy format", () => {
  const rawRecord = {
    stateCode: "US_OZ",
    externalId: "abc123",
    formInformation: {
      clientName: "Firstname Lastname",
      convictionCounty: "Bloom",
      judicialDistrictCode: "XX",
      criminalNumber: "123456",
      judgeName: "Judgename",
      priorCourtDate: "2022-01-01",
      sentenceLengthYears: "2",
      crimeNames: ["Crime1", "crime2"],
      probationExpirationDate: "2024-05-01",
      probationOfficerFullName: "Fakename Officername",
    },
    reasons: [
      {
        criteriaName: "SUPERVISION_EARLY_DISCHARGE_DATE_WITHIN_30_DAYS",
        reason: {
          eligibleDate: "2022-01-03",
        },
      },
      {
        criteriaName: "US_ND_NOT_IN_ACTIVE_REVOCATION_STATUS",
        reason: {
          revocationDate: undefined,
        },
      },
      {
        criteriaName: "US_ND_IMPLIED_VALID_EARLY_TERMINATION_SUPERVISION_LEVEL",
        reason: {
          supervisionLevel: "MEDIUM",
        },
      },
      {
        criteriaName: "US_ND_IMPLIED_VALID_EARLY_TERMINATION_SENTENCE_TYPE",
        reason: {
          supervisionType: "PROBATION",
        },
      },
    ],
    metadata: {
      multipleSentences: false,
      outOfState: false,
      ICOut: false,
    },
  };

  expect(transformReferral(rawRecord)).toMatchSnapshot();
});
