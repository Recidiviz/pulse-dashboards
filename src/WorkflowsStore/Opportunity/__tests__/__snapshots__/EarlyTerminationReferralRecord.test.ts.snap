// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fallback to legacy format 1`] = `
Object {
  "criteria": Object {
    "usNdImpliedValidEarlyTerminationSentenceType": Object {
      "supervisionType": "PROBATION",
    },
    "usNdImpliedValidEarlyTerminationSupervisionLevel": Object {
      "supervisionLevel": "MEDIUM",
    },
    "usNdNotInActiveRevocationStatus": Object {
      "revocationDate": undefined,
    },
  },
  "externalId": "abc123",
  "formInformation": Object {
    "clientName": "Firstname Lastname",
    "convictionCounty": "Bloom",
    "crimeNames": Array [
      "Crime1",
      "crime2",
    ],
    "criminalNumber": "123456",
    "judgeName": "Judgename",
    "judicialDistrictCode": "XX",
    "priorCourtDate": 2022-01-01T00:00:00.000Z,
    "probationExpirationDate": 2024-05-01T00:00:00.000Z,
    "probationOfficerFullName": "Fakename Officername",
    "sentenceLengthYears": 2,
  },
  "metadata": Object {
    "ICOut": false,
    "multipleSentences": false,
    "outOfState": false,
  },
  "stateCode": "US_OZ",
}
`;

exports[`transform record 1`] = `
Object {
  "criteria": Object {
    "supervisionPastEarlyDischargeDate": Object {
      "eligibleDate": 2022-01-03T00:00:00.000Z,
    },
    "usNdImpliedValidEarlyTerminationSentenceType": Object {
      "supervisionType": "PROBATION",
    },
    "usNdImpliedValidEarlyTerminationSupervisionLevel": Object {
      "supervisionLevel": "MEDIUM",
    },
    "usNdNotInActiveRevocationStatus": Object {
      "revocationDate": undefined,
    },
  },
  "externalId": "abc123",
  "formInformation": Object {
    "clientName": "Firstname Lastname",
    "convictionCounty": "Bloom",
    "crimeNames": Array [
      "Crime1",
      "crime2",
    ],
    "criminalNumber": "123456",
    "judgeName": "Judgename",
    "judicialDistrictCode": "XX",
    "priorCourtDate": 2022-01-01T00:00:00.000Z,
    "probationExpirationDate": 2024-05-01T00:00:00.000Z,
    "probationOfficerFullName": "Fakename Officername",
    "sentenceLengthYears": 2,
  },
  "metadata": Object {
    "ICOut": false,
    "multipleSentences": false,
    "outOfState": false,
  },
  "stateCode": "US_OZ",
}
`;
